src/benchmarks/micro/libraries/System.IO.FileSystem/Perf.FileStream.cs

```diff
@@ -2,152 +2,299 @@
             {
-                for (int i = 0; i < TotalSize; i++)
+                for (long i = 0; i < fileSize; i++)
                 {
-                    result += reader.ReadByte();
+                    result += fileStream.ReadByte();
                 }
             }
 
             return result;
         }
-        
+
         [Benchmark]
-        public int Read()
+        [Arguments(OneKiloByte, FileOptions.None)]
+        [Arguments(OneKiloByte, FileOptions.Asynchronous)]
+        public void WriteByte(long fileSize, FileOptions options)
         {
-            byte[] buffer = _buffer;
-            int bytesRead = 0;
-            
-            using (FileStream reader = new FileStream(_filePath, FileMode.Open, FileAccess.Read, FileShare.Read, DefaultBufferSize, FileOptions.None))
+            using (FileStream fileStream = new FileStream(_destinationFilePaths[fileSize], FileMode.Create, FileAccess.Write, FileShare.Read, FourKiloBytes, options))
             {
-                for (int i = 0; i < TotalSize / BufferSize; i++)
+                for (int i = 0; i < fileSize; i++)
                 {
-                    bytesRead += reader.Read(buffer, 0, buffer.Length);
+                    fileStream.WriteByte(default);
                 }
             }
-
-            return bytesRead;
         }
-        
+
+        [GlobalSetup(Targets = new[] { nameof(Read), nameof(ReadAsync), nameof(CopyToFile), nameof(CopyToFileAsync), nameof(Write), nameof(WriteAsync) })]
+        public void SetupBigFileBenchmarks() => Setup(OneKiloByte, OneMegaByte, HundredMegaBytes);
+
         [Benchmark]
-        [BenchmarkCategory(Categories.NoWASM)]
-        public async Task<int> ReadAsync()
+        [Arguments(OneKiloByte, HalfKiloByte, FileOptions.None)] // userBufferSize is less than StreamBufferSize, buffering makes sense
+        [Arguments(OneKiloByte, FourKiloBytes, FileOptions.None)] // the buffer provided by User and internal Stream buffer are of the same size, buffering makes NO sense
+        [Arguments(OneMegaByte, HalfKiloByte, FileOptions.None)]
+        [Arguments(OneMegaByte, FourKiloBytes, FileOptions.None)]
+        [Arguments(HundredMegaBytes, HalfKiloByte, FileOptions.None)]
+        [Arguments(HundredMegaBytes, FourKiloBytes, FileOptions.None)]
+        public long Read(long fileSize, int userBufferSize, FileOptions options)
         {
-            byte[] buffer = _buffer;
-            int bytesRead = 0;
-            
-            using (FileStream reader = new FileStream(_filePath, FileMode.Open, FileAccess.Read, FileShare.Read, DefaultBufferSize, FileOptions.Asynchronous))
+            byte[] userBuffer = _userBuffers[userBufferSize];
+            long bytesRead = 0;
+            using (FileStream fileStream = new FileStream(_sourceFilePaths[fileSize], FileMode.Open, FileAccess.Read, FileShare.Read, FourKiloBytes, options))
```

Comment:
We also end up measuring the open/close time.  That's ok?

Reply:
I would like this benchmark to answer the question "How long does it take to read X bytes from a file?". Since opening (and closing) the file is mandatory to read it, I think that it's OK to include it here.